### PR TITLE
Require seconds if hours and minutes entered

### PR DESCRIPTION
**Pull request by LTA-Thinking**
## Description
Add a requirement that seconds be specified if the hours and minutes in a timestamp are given. This is a replacement of #585.
## Related issues
Addresses #411, #25.

## Testing
Unit tests have been updated to pass seconds and a new one was added to test that seconds are required.

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs [hunk 1]
```diff
@@ -292,7 +301,7 @@ public static PartialDateTime Parse(string s)
             catch (Exception ex) when (ex is ArgumentException)
             {
                 // The input value was parsed correctly but one of the value provided were out of range.
-                throw new FormatException("Input string was not in a correct format. At least one portion of a date was invalid or out of range.", ex);
+                throw new RequestNotValidException("Input string was not in a correct format. At least one portion of a date was invalid or out of range: " + ex.Message);
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: Include ex as the inner exception

File: src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs [hunk 2]
```diff
@@ -292,7 +301,7 @@ public static PartialDateTime Parse(string s)
             catch (Exception ex) when (ex is ArgumentException)
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: Just `catch (ArgumentException ex)`

File: src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs (hunk 1 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: This type of exception doesn't have a constructor that takes an inner exception.

**Comment by johnstairs** [comment]
While you're at it,
``` http
https://localhost:44348/Observation?_lastUpdated=1
```

generates a 500

File: src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs (hunk 1 above)
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: If it do 
``` HTTP
/Observation?_lastUpdated=2013-04-03T15:30
```

This error is returned from the API call:
> Input string was not in a correct format. At least one portion of a date was invalid or out of range: The 'second' portion of a date must be specified if 'hour' and 'minute' are specified.\r\nParameter name: second

The "Parameter name: second" portion does not make sense from the user's context. Ideally, the error would provide the search parameter name, but we don't do that in other cases (but feel free to fix that)

Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: Refactored to remove outer exception.

File: src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs (hunk 2 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: Refactored out the catch

File: src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs (hunk 1 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Core/Models/PartialDateTime.cs]: Fixed exception message

**Comment by LTA-Thinking** [comment]
When the code is parsing times the regex it is using won't match to invalid times. This means that inputs like 2018-02-01T36 or 2018-02-01T12:97 will succeed with the hours and minutes being ignored, while an input like 2018-02-01T12:42:75 will give an exception that seconds is unspecified. Working on remedying this problem.

**Comment by LTA-Thinking** [comment]
Bringing in Issue #25

**Comment by brendankowitz** [comment]
Closing as this PR is covered in #933

**Pull request closed without merging**